### PR TITLE
[BugFix] Record query result if forward to leader

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes [#1081](https://github.com/StarRocks/StarRocksTest/issues/1081)

## Problem Summary(Required) ：

Fixed a legacy bug that existed long before I joined the company.

Sometimes a SQL received by a follower FE may end up forwarding to the leader. The leader will process the statement and send the result back to the follower FE. The result will be stored in the `ShowResultSet` field of the thrift response message sent from the master to the follower. 

However, if the forwarded statement is a query statement, the query result has no place to store in the response and will be directly written to the send buffer of the leader MySQL channel. After the follower receives the response message from the leader, it will find out that the `ShowResultSet` field is empty and will regard the master as a lower version of FE; thus, it will send a simple EOF to the MySQL client. The MySQL client receives an unexpected EOF and will report a `malformated packet` error.

This bug is seldom triggered because we never forward a query statement unless when the progress of the follower is far behind the leader. Even in that scenario, we would try to fix the FE so that the follower can catch up with the leader. As a result, this bug is ignored.

Luckily, in developing materialized view, we rewrite the `show materialized` statement to a query in the `information_schema.materialized_views` table. This bug has finally revealed its veil.

To fix this bug, we need to add a list of Bytebuffer in the thrift response message from the leader to the master. All the packets that are supposed to send to the MySQL channel of the leader will be stored in this list so that they can send to the MySQL client after the follower receives the response.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/LeaderOpExecutor.java [hunk 1]
```diff
@@ -163,6 +165,20 @@ public ShowResultSet getProxyResultSet() {
         }
     }
 
+    /**
+     * if a query statement is forwarded to the master, or if a show statement is automatically rewrote,
```
Comment (waittttting) [inline fe/fe-core/src/main/java/com/starrocks/qe/LeaderOpExecutor.java:169]: We not use master anymore

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/qe/LeaderOpExecutor.java]: Good point. Fixed it.

**Review by waittttting — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 8 / 37 (21.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/LeaderOpExecutor.java|0|6|00.00%|[173, 174, 176, 177, 178, 179]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|0|11|00.00%|[492, 493, 494, 495, 496, 499, 646, 647, 648, 649, 650]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|20|40.00%|[181, 294, 295, 297, 749, 759, 760, 762, 966, 975, 985, 1522]|

**Comment by padmejin** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java [hunk 2]
```diff
@@ -489,8 +489,16 @@ private void finalizeCommand() throws IOException {
                 // ShowResultSet is null means this is not ShowStmt, use remote packet(executor.getOutputPacket())
                 // or use local packet (getResultPacket())
                 if (resultSet == null) {
-                    packet = executor.getOutputPacket();
-                } else {
+                    if (executor.sendResultToChannel(ctx.getMysqlChannel())) {  // query statement result
+                        packet = getResultPacket();
+                        if (packet == null) {
+                            LOG.debug("packet == null");
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java:495]: What's the case?

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/qe/ConnectProcessor.java]: Just removed these codes. Thanks for your reminder.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 8 / 34 (23.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/LeaderOpExecutor.java|0|6|00.00%|[173, 174, 176, 177, 178, 179]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|0|8|00.00%|[492, 493, 495, 642, 643, 644, 645, 646]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|8|20|40.00%|[181, 294, 295, 297, 749, 759, 760, 762, 966, 975, 985, 1522]|

**Review by gengjun-git — APPROVED** [review]

**Comment by padmejin** [comment]
run starrocks_admit_test

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
@Mergifyio backport branch-2.4
¡Muchas gracias!

**Pull request merged**